[NEW] nginx-ingress

mgoodness opened the pull request:
This is a stand-alone chart for the nginx ingress controller that, along with the kube-lego chart (#397), replaces nginx-lego. As mentioned and (briefly) discussed in #385.

@jackzampolin

Commit:
Renamed nginx-lego to nginx-ingress

* Updated common labels (app, heritage, release)
* Added component label per proposed best practices
* Changed container names per proposed best practices
* Added default-backend livenessProbe
* Updated default-backend to v1.2
* Replaced individual nginx.configmap values with toYaml
* Added livenessProbe, readinessProbe, and port names to nginx
Deployment
* Enabled creation of multiple nginx Services
* Set nginx resources to match upstream
* Truncate templates at 63 chars

Commit:
nginx-ingress v0.2.8

* Restructure

Commit:
NOTES & README update

jackzampolin commented:
@mgoodness This looks great! I'd be happy to stay on as the maintainer! 😉  It's much better as two charts. I'll pull this down and give it a test.

Commit:
Updated NOTES.txt

viglesiasce commented:
Tested this over the weekend without any issues.

Will wait for @jackzampolin to give the LGTM

bacongobbler commented:
ping @jackzampolin, does this have your LGTM?

jackzampolin commented:
@bacongobbler Got to this today. LGTM. Sorry for the delay there.

prydonius commented:
This is the first time we're changing the name/deprecating a chart, since nginx-lego will stay in the repo, is there something we can do to let people know it's deprecated? e.g. we could push a final version that explains it's moved to nginx-ingress in the NOTES.txt?

jackzampolin commented:
@prydonius I don't know how many users it had. In this particular case there was an issue (my b) with the other chart that prevented it from deploying successfully. 

It would be nice to do in the future but I think it might be overkill for this particular instance.

jason-riddle commented:
I ran into a small hiccup in regards to the instructions.

```
export NODE_PORT=$(kubectl --namespace default get services -o jsonpath="{.spec.ports[0].nodePort}" silly-aardvark-nginx-ingress-controller)

export NODE_IP=$(kubectl --namespace default get nodes -o jsonpath="{.items[0].status.addresses[0].address}")

echo "Visit http://$NODE_IP:$NODE_PORT to access your application."
```

~~~I examined the NODE_IP which appears to be using the internal cluster ip. I believe this is due to the fact that I applied this chart with~~~

```
  services:
    - name: controller
      # annotations:
      type: NodePort
```

~~So after I created my ingress, I queried it and it appear to has the correct external facing ip.~~

Update below.

jason-riddle commented:
Update: I mispoke. The problem for me was 

```
kubectl --namespace default get nodes -o jsonpath="{.items[0].status.addresses[0].address}"
```

should be

```
kubectl --namespace default get nodes -o jsonpath="{.items[0].status.addresses[1].address}"
```

so

```bash
status.addresses[1]
# instead of status.addresses[0]
```

prydonius commented:
@jason-riddle it appears that a lot of charts use this in their NOTES, can you create a separate issue for it?

jason-riddle commented:
sure

munnerz commented:
Hey @mgoodness - I've just opened https://github.com/kubernetes/charts/pull/540, and then gone on to spot this PR :)

You seem to have put more into documenting so think it's best we stick with your PR, but may I request you add support for TCP & UDP configmaps into yours too? You can see an example of how I've done it in the above linked PR. I'm happy to do it myself, but will need to wait until after this has merged if so!

viglesiasce commented:
@k8s-bot e2e test this

viglesiasce commented on stable/nginx-ingress/templates/controller-service.yaml:
```diff
@@ -0,0 +1,29 @@
+{{- $root := . -}}
+{{- range .Values.controller.services }}
+---
```
Tests are failing because of this. Do we have to loop in this file? When would I have multiple? Can we start with this hardcoded to 1 service?

mgoodness replied:
My specific use-case is to have the same nginx-ingress controllers exposed via two different ELBs (one internet-facing and one VPC-internal). But now that I think about it, that shouldn't be strictly necessary. I'll play with it over the weekend.

Commit:
v0.3.0

* expose single service
* enable TCP & UDP load-balancing

mgoodness commented:
Addressed comments & requests from @viglesiasce, @jason-riddle, and @munnerz.

Commit:
install as Deployment or DaemonSet

kfox1111 commented on stable/nginx-ingress/templates/controller-daemonset.yaml:
```diff
@@ -0,0 +1,76 @@
+        app: {{ template "name" . }}
+        component: "{{ .Values.controller.name }}"
+        release: {{ .Release.Name }}
+    spec:
```
Needs an optional nodeSelector section from values.

Otherwise, it may land ingress controllers on unsuitable nodes.

kfox1111 commented on stable/nginx-ingress/templates/controller-deployment.yaml:
```diff
@@ -0,0 +1,77 @@
+        app: {{ template "name" . }}
+        component: "{{ .Values.controller.name }}"
+        release: {{ .Release.Name }}
+    spec:
```
Optional nodeSelector.

kfox1111 commented on stable/nginx-lego/templates/default-deployment.yaml:
```diff
@@ -1,21 +0,0 @@
-  name: {{ template "fullname" . }}-default-backend
-  labels:
-    chart: "{{ .Chart.Name }}-{{ .Chart.Version }}"
-spec:
```
nodeSelector

Could this be added as a sidecar to help monitor easier?
https://github.com/hnlq715/nginx-vts-exporter

mgoodness replied:
This file is being deleted.

kfox1111 replied:
Doh, your right. Sorry.

What do you think of the nginx-vtx-exporter sidecar though? Still applies to the new version too.

kfox1111 commented on stable/nginx-ingress/templates/nginx-deployment.yaml:
```diff
@@ -0,0 +1,58 @@
+        component: controller
+        release: "{{ .Release.Name }}"
+    spec:
+      containers:
```
Could you add this as a sidecar to make monitoring easier?
https://github.com/hnlq715/nginx-vts-exporter

kfox1111 commented on stable/nginx-lego/templates/nginx-deployment.yaml:
```diff
@@ -1,38 +0,0 @@
-      labels:
-        app: {{ template "fullname" . }}
-    spec:
-      containers:
```
https://github.com/hnlq715/nginx-vts-exporter 

nodeSelector

mgoodness replied:
This file is being deleted.

kfox1111 commented on stable/nginx-ingress/templates/default-deployment.yaml:
```diff
@@ -0,0 +1,35 @@
+        app: {{ template "fullname" . }}
+        component: default-backend
+        release: "{{ .Release.Name }}"
+    spec:
```
Needs an optional nodeSelector to restrict where it lands in non uniform clusters.

kfox1111 commented on stable/nginx-ingress/templates/nginx-service.yaml:
```diff
@@ -0,0 +1,26 @@
+    heritage: {{ $root.Release.Service }}
+    release: {{ $root.Release.Name }}
+  name: "{{ template "fullname" $root }}-controller-{{ .name }}"
+spec:
```
This needs an optional externalIPs section. This would allow it to easily be able to be bound to vips for clusters that use those.

kfox1111 reviewed (changes requested):
This looks really good. I went to test it on my cluster but ran into a few small issues preventing deployment in my environment. Specifically with not being able to pin it to some node types and the lack of externalIPs on the main service.

unguiculus commented on stable/nginx-ingress/templates/controller-service.yaml:
```diff
@@ -0,0 +1,46 @@
+    release: {{ .Release.Name }}
+  name: {{ template "fullname" . }}-{{ .Values.controller.name }}
+spec:
+{{- if .Values.loadBalancerIP }}
```
Should be `.Values.service.loadBalancerIP`

unguiculus commented on stable/nginx-ingress/templates/controller-service.yaml:
```diff
@@ -0,0 +1,46 @@
+{{- if .Values.loadBalancerIP }}
+  loadBalancerIP: {{ .Values.service.loadBalancerIP }}
+{{- end }}
+{{- if .Values.loadBalancerSourceRanges }}
```
Should be `.Values.service.loadBalancerSourceRanges`

unguiculus commented on stable/nginx-ingress/templates/controller-service.yaml:
```diff
@@ -0,0 +1,46 @@
+{{- end }}
+{{- if .Values.loadBalancerSourceRanges }}
+  loadBalancerSourceRanges:
+  {{- range $cidr := .Values.loadBalancerSourceRanges }}
```
Should be `.Values.service.loadBalancerSourceRanges`

kfox1111 commented on stable/nginx-ingress/templates/default-backend-deployment.yaml:
```diff
@@ -0,0 +1,35 @@
+apiVersion: extensions/v1beta1
```
Could you make this, and the default-backend-svc.yaml optional the same way the lego stuff was?

This would allow a user to load in their own default server, like a regular nginx to serve out a site customized/branded 404 page.

mgoodness replied:
A user could do that using the current setup. Just need to set the `defaultBackend` values appropriately.

kfox1111 replied:
Oh, I guess you could load a different image using the defaultBackend properties. So that might work for some use cases.

It would not give you the ability to load a differently setup pod though. So, like, if I wanted to load a deployment with an nginx container for my branded page, a fluentd sidecar to pull logs, and a prometheus exporter to monitor it.

So I still think the ability to disable its automatic creation is valuable?

Commit:
PR work

* default backend can be user-provided
* added nodeSelectors
* vts-nginx-exporter
* service fixes

mgoodness commented:
Not a huge fan of how I handled the default backend, so I'm open to suggestions. Will touch up the README tomorrow.

kfox1111 commented:
Thanks for fixing the lack of nodeSelector and adding the exporter sidecar.

I tested it out and its mostly working except I had to add the following to the controller-daemonset and deployment:

      annotations:
        "prometheus.io/scrape": {{ .Values.controller.enableStats | quote }}
        "prometheus.io/port": {{ .Values.statsExporter.service.containerPort | quote }}

A Prometheus instance I launched from the chart in this repo then automatically picked it up and started collecting metrics on all the instances without needing any configuration. Very cool. :)

Could you please add that bit to the pr?

Commit:
Documentation & PR

* added podAnnotations
* updated README

mgoodness commented:
Prometheus annotations (or any others) can now be listed in the `podAnnotations` values:
```
controller:
  podAnnotations:
    prometheus.io/scrape: "true"
    prometheus.io/port: "9913"
    ...
```
While these annotations are the ones recommended by Prometheus, they are not universal or mandatory. Users may use different ones for service discovery. I'm opting for flexibility.

kfox1111 commented:
Got a chance to play with this a bit more tonight.

I still need a way to set:
  externalIPs:
    - xxx.xxx.xxx.xxx
    - yyy.yyy.yyy.yyy
on the services.

Once I did that, I was able to bind the ingress to Prometheus and see its ui outside my baremetal cluster.

Commit:
added externalIPs to controller & stats exporter services

mgoodness commented:
@k8s-bot e2e test this

Commit:
Additional configurability

- extraArgs for all containers
- annotations, clusterIP, externalIPs, loadBalancerIP, loadBalancerSourceRanges for all services
- separate controller.stats section

mgoodness commented:
@kfox1111 have you had a chance to test the latest iteration? I think it's good to go.

Commit:
Fixed metrics servicePort

Commit:
Use templates for fullnames

Commit:
Renamed enableStats -> stats.enabled

Commit:
Deprecate nginx-lego, not delete

mgoodness commented:
After further discussion with @prydonius, we've decided not to delete nginx-lego as part of this PR. Instead, I'll create another in which the README and NOTES.txt make it clear that it's been deprecated. There's also [work](https://github.com/kubernetes/helm/issues/1963) planned in the Helm project to support/enforce deprecation.

prydonius commented on stable/nginx-ingress/Chart.yaml:
```diff
@@ -0,0 +1,14 @@
+name: nginx-ingress
+version: 0.3.0
+description: An nginx Ingress controller that uses ConfigMap to store the nginx configuration.
+keywords:
```
Can we add an icon here? e.g. pointing to https://cdn.wp.nginx.com/wp-content/themes/nginx-theme/assets/img/logo.png?

Commit:
Added icon

Commit:
Use nginx.org logo

Commit:
No default resources

prydonius commented on stable/nginx-ingress/README.md:
```diff
@@ -0,0 +1,118 @@
+`controller.stats.service.externalIPs` | controller service stats external IP addresses | `[]`
+`controller.stats.service.loadBalancerIP` | IP address to assign to load balancer (if supported) | `""`
+`controller.stats.service.loadBalancerSourceRanges` | list of IP CIDRs allowed access to load balancer (if supported) | `[]`
+`controller.service.type` | type of controller stats service to create | `ClusterIP`
```
should this be `controller.stats.service.type`?

prydonius commented on stable/nginx-ingress/README.md:
```diff
@@ -0,0 +1,118 @@
+`defaultBackend.service.externalIPs` | default backend service external IP addresses | `[]`
+`defaultBackend.service.loadBalancerIP` | IP address to assign to load balancer (if supported) | `""`
+`defaultBackend.service.loadBalancerSourceRanges` | list of IP CIDRs allowed access to load balancer (if supported) | `[]`
+`controller.service.type` | type of default backend service to create | `ClusterIP`
```
defaultBackend.service.type?

prydonius commented:
@mgoodness just some small fixes in the README, otherwise this lgtm!

Commit:
PR feedback
